plutus-core-interpreter/src/Language/PlutusCore/Interpreter/LMachine.hs

```diff
@@ -0,0 +1,278 @@
+
+lookupHeap :: HeapLoc -> Heap -> HeapEntry
+lookupHeap l (Heap h _) =
+    case IntMap.lookup l h of
+      Just e  -> e
+      Nothing -> throw $ LMachineStringException ("Missing heap location in lookupHeap: " ++ show l)  -- This should never happen
+
+
+-- | The basic computation step of the L machine.  Search down the AST looking for a value, saving surrounding contexts on the stack.
+computeL :: EvaluationContext -> Heap -> Closure -> LMachineResult
+computeL ctx heap cl@(Closure term env) =
+    case term of
+      TyInst _ fun ty       -> computeL (FrameTyInstArg ty : ctx)                 heap (Closure fun env)
+      Apply _ fun arg       -> computeL (FrameAppArg (Closure arg env) : ctx)     heap (Closure fun env)
+      Wrap ann tyn ty term' -> computeL (FrameWrap ann tyn ty : ctx)              heap (Closure term' env)
+      Unwrap _ term'        -> computeL (FrameUnwrap : ctx)                       heap (Closure term' env)
+      TyAbs{}               -> returnL  ctx heap cl
+      LamAbs{}              -> returnL  ctx heap cl
+      Builtin{}             -> returnL  ctx heap cl
+      Constant{}            -> returnL  ctx heap cl
+      Error{}               -> Failure
+      Var _ name            -> let l = lookupHeapLoc name env
+                               in case lookupHeap l heap of
+                                    Evaluated cl'   -> returnL ctx heap cl'
+                                    Unevaluated cl' -> computeL (FrameHeapUpdate l : ctx) heap cl'
+
+
+-- | Return a closure containing a value. Ideally the fact that we've
+-- got a value would be enforced in the (Haskell) type. Values still
+-- have to be contained in closures.  We could have something like
+-- @v= \x.y@, where @y@ is bound in an environment.  If we ever go on
+-- to apply @v@, we'll need the value of @y@.
+returnL :: EvaluationContext -> Heap -> Closure -> LMachineResult
+returnL [] heap res                                                        = Success res heap
+returnL (FrameTyInstArg ty        : ctx) heap cl                           = instantiateEvaluate ctx heap ty cl
+returnL (FrameAppArg argClosure   : ctx) heap cl                           = evaluateFun ctx heap cl argClosure
+returnL (FrameWrap ann tyn ty     : ctx) heap (Closure v env)              = returnL ctx heap $ Closure (Wrap ann tyn ty v) env
+returnL (FrameHeapUpdate l        : ctx) heap cl                           = returnL ctx heap' cl
+                                                                                   where heap' = updateHeap l (Evaluated cl) heap
+                                                                                      -- Leave this out to make the machine call-by-name (really slow)
+returnL (FrameUnwrap              : ctx) heap (Closure (Wrap _ _ _ t) env) = returnL ctx heap (Closure t env)
+returnL (FrameUnwrap              : _)      _ (Closure term _)             = throw $ LMachineException "Attemtping to unwrap non-wrapped term" term
```

Comment:
I'd prefer to avoid very long strings. This one is 147 symbols long.

Reply:
You're probably right.  There's an uncomfortable conflict between getting things nicely aligned and making things too long.

Reply:
I changed it to do a case split on the frames and that makes it a lot more compact.